llvm/lib/CodeGen/Analysis.cpp

```diff
@@ -685,7 +687,10 @@ bool llvm::returnTypeIsEligibleForTailCall(const Function *F,
          (IID == Intrinsic::memmove &&
           TLI.getLibcallName(RTLIB::MEMMOVE) == StringRef("memmove")) ||
          (IID == Intrinsic::memset &&
-          TLI.getLibcallName(RTLIB::MEMSET) == StringRef("memset"))) &&
+          TLI.getLibcallName(RTLIB::MEMSET) == StringRef("memset") &&
+          (!isa<ConstantInt>(Call->getOperand(1)) ||
+           !cast<ConstantInt>(Call->getOperand(1))->isZero() ||
+           !TLI.getLibcallName(RTLIB::BZERO)))) &&
```

Comment:
This is sort of ugly... if this code doesn't precisely match SelectionDAG::getMemset, we miscompile.

Can we rearrange the code so that the caller of isInTailCallPosition() does this computation?  That could be directly adjacent to the code in getMemset that actually makes the decision of whether to call memset or bzero.  We then just pass that into isInTailCallPosition() as a argument.

Reply:
Maybe I'm not understanding what you're suggesting. `isInTailCallPosition` is called in multiple places so adding in an extra parameter doesn't make it any cleaner IMO. getMemset() is called later, by that point we've already decided to tail-call. We could downgrade the tail call in getMemset() but then we also lose the tail-call when the parent function's return type was void or it returned undef.

Reply:
Currently, isTailCall is a parameter to SelectionDAG::getMemset(), but it doesn't really need to be; SelectionDAGBuilder::visitIntrinsicCall just calls isInTailCallPosition so it can pass the result into getMemset().  Instead, we can just call isInTailCallPosition from getMemset(), once we know whether we're calling bzero or memset, and we can pass that information down.

Reply:
I tried to implement this. It didn't turn out as well as I'd hoped, memset is ok to do and we the logic for memset. However for memmove we have a user in XCore which is computing a tail-call-position using a DAG node without any reference to a `CallInst` so we can't do the same thing there. From what I could tell memcpy usages seemed ok to convert too but I didn't check all of them since there were so many.

Reply:
Due to the way calls are lowered, we have two different tail-call-position analysis passes, yes: one for regular calls, which are lowered as part of building the SelectionDAG, and one for library calls which are created as part of legalization/optimization/etc.

We could teach getMemmove() etc. to call the correct isInTailCallPosition() based on the arguments: if the caller passed in an Instruction, use the IR version, if the caller passed in an SDNode, use the SelectionDAG version.  Or we could try to use the SelectionDAG version exclusively: add an ISD::MEMMOVE node, and lower it after SelectionDAGBuilder.

Reply:
I think only the IR version tries to do the returns-first-arg trickery so we can probably just go with the first option.